File: 第四章_队列和栈/队列和栈(3)/滑动窗口练习题/SlideWindow.cpp
class SlideWindow {
public:
    vector<int> slide(vector<int> arr, int n, int w) {
        vector<int> res;
        deque<int> dq;
        for(int i=0;i<n;i++){
            if(dq.empty()||arr[dq.back()]>arr[i])//当队列为空或者队尾索引下的元素大于数组元素，则入队索引
                dq.push_back(i);
            else{
                while(!dq.empty()&&arr[dq.back()]<arr[i]){//否则就一直出队
                    dq.pop_back();
                }
                dq.push_back(i);
            }
            if(i>=w-1){//赋值最大值数组
                while(!dq.empty() && i-dq.front()>w-1)
                    dq.pop_front();
                res.push_back(arr[dq.front()]);
            }
        }
        return res;
    }
};

File: 第四章_队列和栈/队列和栈(3)/滑动窗口练习题/SlideWindow.java
import java.util.ArrayList;

public class SlideWindow {
	public int[] slide(int[] arr, int n, int w) {
		ArrayList<Integer> list = new ArrayList<>();// 双端队列 保存数组下标索引
		int[] res = new int[n - w + 1];// 存放最大值结果数组

		for (int i = 0; i < n; i++) {
			if (!list.isEmpty() && list.get(0) <= i - w) {// 当队列的头元素超出了当前滑动窗口的范围时，就移除。
				list.remove(0);
			}

			while (!list.isEmpty() && arr[list.get(list.size() - 1)] < arr[i]) {// 当前队列的队尾元素的值小于数组值，移除队尾元素
				list.remove(list.size() - 1);
			}

			if (list.isEmpty() || arr[list.get(list.size() - 1)] > arr[i]) {// 队尾元素大于数组元素，加入当前索引
				list.add(i);
			}

			if (i >= w - 1) {// 赋值最大值
				res[i - w + 1] = arr[list.get(0)];
			}
		}

		return res;
	}
}